protocol/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,77 @@
+
+#### Type Hierarchy
+
+Many language support inheritance and LSP provides retrieving the hierarchy information with the following requests.
+
+
+The `textDocument/subTypes` request is sent from the client to the server to resolve resolve the subtypes of a symbol at a given text document position. Returns the direct subtypes in no particular order.
+
+_Client Capabilities_:
+
+```ts
+  TextDocumentClientCapabilities {
+    /**
+     * Capabilities specific to the `textDocument/subTypes`
+     *  and `textDocument/superTypes
+     */
+    typeHierarchy?: {
+    /**
+     * Whether implementation supports dynamic registration.
+     */
+    dynamicRegistration?: boolean;
+  };
+```
+
+_Server Capabilities_:
+```ts
+ServerCapabilities{
+  /**
+   * The server provides type hierarchy information
+   */
+  typeHierarchyProvider?: boolean
+}
+```
+
+```ts
+/**
+ * Represents hierarchical information about programming constructs like variables, classes,
+ * interfaces etc.
+ */
+export interface SymbolNode extends SymbolInformation {
+  /**
+   * Immediate descendants of this node
+   */
+  descendants? : SymbolNode[]
+  /**
+   * true if this node has children. hasChildren can be
+   * true even when the descendants are empty. In that case
+   * the clients should do a call to retrieve the descendants.
+   */
+  hasChildren: boolean
+}
+```
+
+_Request_:
+
+method: ‘textDocument/subTypes’
+params: TextDocumentPositionParams
+
+_Response_
+
+* result: `SymbolNode[] | null`
+* error: code and message set in case an exception happens during the 'textDocument/subTypes' request
```

Comment:
When I open a type hierarchy on a certain position in the text, I would like to see the `SymbolNode` for that token together with its subtypes. How do I obtain the SymbolNode I am opening the hierarchy on?

Reply:
There can be multiple dimensions of sub-super relations: template parameters related to a template, derived types related to a base, specializations. Is there some way to differentiate them in this proposal?